Docs updates (#28)
flare-foundation/attestation-client: docs/indexing/indexer-optimizations.md

- Get block by block number or hash
- Get block height

> Caching of API responses is important mainly for transactions. An important use case is Bitcoin and other Bitcoin code based UTXO blockchains. Specifically, in UTXO blockchains a source address is obtained by querying the blockchain API for each transaction on each input. There can be up to 20 inputs and possibly many different UTXO outputs from different transactions. Usually we can get a block info and info of all transactions in the block in one request. But then we have to make many requests for input transactions for each of those transactions in the block. Those could repeat and caching responses helps here. Note that some blocks can have over 2000 transactions and some experimental calculations showed that we would for each transaction do 4 additional API calls on average for input transactions. With Bitcoin blocks of over 2000 transactions this may take even 5 minutes. Compared to 10 minute average block production of Bitcoin the indexing would take quite some time. Also, the most critical requirement for an indexer is that confirmed transactions are stored into the database immediately after the confirmation block arrives. Hence we have to process transactions from unconfirmed blocks in advance. Since unconfirmed blocks may be in different forks, we might be processing wrong fork. But even if we are in the wrong fork, a lot of transactions in that wrong block will appear in the right block as well. So here caching can also help greatly.

Reviewer: ```suggestion
Caching of API responses is important mainly for transactions. An important use case is Bitcoin and other Bitcoin-based UTXO blockchains. Specifically, in UTXO blockchains a source address is obtained by querying the blockchain API for each transaction on each input. There can be up to 20 inputs and possibly many different UTXO outputs from different transactions. Usually we can get a block info and info of all transactions in the block in one request. But then we have to make many requests for input transactions for each of those transactions in the block. Those could repeat and caching responses helps here. Note that some blocks can have over 2000 transactions and experimental calculations showed that for each transaction we would do 4 additional API calls on average for input transactions. With Bitcoin blocks of over 2000 transactions this may even take 5 minutes. Compared to the average 10-minute block production of Bitcoin, the indexing would take quite some time. Also, the most critical requirement for an indexer is that confirmed transactions are stored into the database immediately after the confirmation block arrives. Hence we have to process transactions from unconfirmed blocks in advance. Since unconfirmed blocks may be in different forks, we might be processing the wrong fork. But even if we are in the wrong fork, a lot of transactions in that wrong block will appear in the right block as well, so caching can still help greatly.
```

> We also note that we used another optimization: we only read all inputs of transactions that have payment reference defined. For others it depends on attestation type, whether we do additional reads during verification phase or not.

Reviewer: ```suggestion
We also note that we used another optimization: we only read all inputs of transactions that define a payment reference. For other transactions, whether we do additional reads during verification phase or not depends on the attestation type.
```
